server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTarget.java

```diff
@@ -177,14 +188,53 @@ public void startReplication(ActionListener<Void> listener) {
     private List<StoreFileMetadata> getFiles(CheckpointInfoResponse checkpointInfo) throws IOException {
         cancellableThreads.checkForCancel();
         state.setStage(SegmentReplicationState.Stage.FILE_DIFF);
-        final Store.RecoveryDiff diff = Store.segmentReplicationDiff(checkpointInfo.getMetadataMap(), indexShard.getSegmentMetadataMap());
-        logger.trace(() -> new ParameterizedMessage("Replication diff for checkpoint {} {}", checkpointInfo.getCheckpoint(), diff));
+        Map<String, StoreFileMetadata> activeReaderMetadataMap = indexShard.getSegmentMetadataMap();
+        Map<String, StoreFileMetadata> primaryMetadataMap = checkpointInfo.getMetadataMap();
+
+        Store.RecoveryDiff diff = Store.segmentReplicationDiff(primaryMetadataMap, activeReaderMetadataMap);
+        List<String> unreferencedDiskFiles = Arrays.asList(indexShard.store().directory().listAll())
```

Comment:
if you loop over diff.missing first you can identify the specific files if they are also on disk. This would exclude segments_n and excluded lock files already.

Example:

```java
        // local files
        final List<String> localFiles = List.of(store.directory().listAll());
        // set of local files that can be reused
        final Set<StoreFileMetadata> reuseFiles = diff.missing.stream()
            .filter(storeFileMetadata -> localFiles.contains(storeFileMetadata.name()))
            .filter(this::validateLocalChecksum)
            .collect(Collectors.toSet());
```

with validateLocalChecksum something like

```java
    private boolean validateLocalChecksum(StoreFileMetadata file) {
        try (IndexInput indexInput = indexShard.store().directory().openInput(file.name(), IOContext.DEFAULT)) {
            String checksum = Store.digestToString(CodecUtil.retrieveChecksum(indexInput));
            if (file.checksum().equals(checksum)) {
                return true;
            } else {
                // clear local copy with mismatch. Safe because file is not referenced by active reader.
                store.deleteQuiet(file.name());
                return false;
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
```

Reply:
Thanks for suggesting this optimization. This looks pretty clean!